extension_service/datastore/providers/firestore.py

```diff
@@ -0,0 +1,270 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from datetime import datetime, timedelta
+from typing import Any, Dict, Literal, Optional
+
+import firebase_admin
+from firebase_admin import credentials, firestore_async
+from google.cloud import firestore
+from google.cloud.firestore_v1.base_query import FieldFilter
+from pydantic import BaseModel
+
+import models
+
+from .. import datastore
+
+
+class Config(BaseModel, datastore.AbstractConfig):
+    kind: Literal["firestore"]
+    projectId: str
+    serviceAccountId: str
+
+
+class Client(datastore.Client[Config]):
+    __client: firestore.AsyncClient
+
+    @datastore.classproperty
+    def kind(cls):
+        return "firestore"
+
+    def __init__(self, client: firestore.AsyncClient):
+        self.__client = client
+
+    @classmethod
+    async def create(cls, config: Config) -> "Client":
+        cred = credentials.ApplicationDefault()
+        firebase_admin.initialize_app(
+            cred,
+            options={
+                "projectId": config.projectId,
+                "serviceAccountId": config.serviceAccountId,
+            },
+        )
+        return cls(firestore.AsyncClient(project=config.projectId))
+
+    async def initialize_data(
+        self,
+        airports: list[models.Airport],
+        amenities: list[models.Amenity],
+        flights: list[models.Flight],
+    ) -> None:
+        async def delete_collection(coll_ref, batch_size=400):
+            docs = coll_ref.limit(batch_size).stream()
+            deleted = 0
+
+            async for doc in docs:
+                await doc.reference.delete()
+                deleted = deleted + 1
+
+            if deleted >= batch_size:
+                return await delete_collection(coll_ref, batch_size)
+
+        # Check if the collections already exist; if so, delete collections
+        airports_ref = self.__client.collection("airports")
+        airports_exist = await airports_ref.limit(1).get()
+        if airports_exist:
+            await delete_collection(airports_ref)
+
+        amenities_ref = self.__client.collection("amenities")
+        amenities_exist = await amenities_ref.limit(1).get()
+        if amenities_exist:
+            await delete_collection(amenities_ref)
+
+        flights_ref = self.__client.collection("flights")
+        flights_exist = await flights_ref.limit(1).get()
+        if flights_exist:
+            await delete_collection(flights_ref)
```

Comment:
Any reason we shouldn't try to await these together -- and get some benefit from async here?

Reply:
Updated: we will await all the delete together.